KAFKA-5488: Add type-safe split() operator

inponomarev opened the pull request:
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

inponomarev committed:
public interfaces and JavaDoc

inponomarev committed:
added implementation class

inponomarev committed:
update reference in JavaDoc

inponomarev committed:
remove compiler warnings

inponomarev committed:
implementation and test

inponomarev committed:
more tests

inponomarev committed:
Documentation update

inponomarev commented:
⚠️ Two differences with KIP specification, discussion needed⚠️ 

1. Instead of multiple overloaded variants of `Branched.with` we now have `Branched.withFunction` and `Branched.withConsumer`. This is because of compiler warnings about overloading (`Function` and `Consumer` being indistinguishable when supplied as lambdas)

2. 'Fully covariant' signatures like `Consumer<? super KStream<? super K, ? super V>>` don't work as expected. Used `Consumer<? super KStream<K, V>>` instead

vvcephei commented:
Hey @mjsax , do you have time to give this a first pass?

mjsax commented:
I'll put it into my backlog. But I am the main reviewer for two other KIPs (216 and 466) that I should review first as they got approve earlier and PRs are open for longer already.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,150 @@
+
+    protected final String name;
+    protected final Function<? super KStream<K, V>,
+            ? extends KStream<K, V>> chainFunction;
```
nit: should be one line (easier to read)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    /**
+     * Create an instance of {@link Branched} from an existing instance.
```
nit: `{@link ' -> `{@code` (we should use `@link` to link to other classes but use `@code` to refer to ourself -- there is no point in letting a JavaDoc page link to itself.

Similar below.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,150 @@
+     * @return {@code this}
+     */
+    @Override
+    public Branched<K, V> withName(final String name) {
```
nit: can you first have all the static builder methods and the instance methods at the end of the class?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,150 @@
+     * Create an instance of {@link Branched} with provided chain function.
+     *
+     * @param chain A function that will be applied to the branch. If {@code null}, the identity
+     *              {@code kStream -> kStream} function will be supposed. If this function returns
```
nit `supposed` -> `used` ?

`this function` -> `the provided function`

mjsax replied:
Actually, I am wondering if we should allow to pass in `null`? Thoughts?

inponomarev replied:
See my reply below, where we discuss `null` consumers: https://github.com/apache/kafka/pull/9107#discussion_r548515181

(in short: I agree, I think we shouldn't)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,150 @@
+    /**
+     * Create an instance of {@link Branched} with provided chain consumer.
+     *
+     * @param chain A consumer to which the branch will be sent. If a non-null branch is provided here,
```
`If a non-null branch is provided here`? (`branch` -> `consumer`?)

But I would propose to simplify it, and just use `By default` (as passing in a non-null consumer should be the "default" usage).

mjsax replied:
As above, should we even allow a not-null consumer?

inponomarev replied:
see https://github.com/apache/kafka/pull/9107#discussion_r548515181

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,150 @@
+     * @param chain A consumer to which the branch will be sent. If a non-null branch is provided here,
+     *              the respective branch will not be added to the resulting {@code Map} returned
+     *              by {@link BranchedKStream#defaultBranch()} or {@link BranchedKStream#noDefaultBranch()} (see
+     *              {@link BranchedKStream} description for details). If {@code null}, a no-op consumer will be supposed
```
`If {@code null} is passed instead of an actual consumer, the branch will be added unmodified to the resulting {@code Map}.`

mjsax replied:
I tried to read up on the KIP discussion thread, and I am wondering if we did agree to this behavior? My understanding was that if a consumer is use, there won't be any entry in the `Map` for this branch?

inponomarev replied:
OK let me explain how I understood this: 

Let's try to forget about Java type system and look at the problem conceptually, from set / category theory point of view

We're using either **function** or **consumer** or **nothing** for branch processing. From mathematical point of view, we can consider consumer to be a function that maps its domain to the empty set. This explains our decision why consumer 'swallows' branch unlike function that 'forwards' its result to the resulting map. Then we agreed that the case when **nothing** is provided, like in this example

```java
BranchedKStream<Integer, String> branch = source.split()
          .branch(isEven)
          .branch(isMultipleOfSeven);
```
is to be interpreted like 

```java
BranchedKStream<Integer, String> branch = source.split()
         .branch(isEven, Branched.withFunction(Function.identity()))
         .branch(isMultipleOfSeven, Branched.withFunction(Function.identity()));
```

Now back to Java: in Java, for `nothing` we have `null`. My original intention was to try to emulate `Function|Consumer` union type using overloading, but Java didn't allow me to do this dirty trick -- and separate `withFunction` and `withConsumer` builder methods appeared 😄 

If I had succeded, it would have made sence to treat `null` uniformely for both functions and consumers. 

Hope this explains why I allowed `null`s and always treated it as an identity function.

But now I think you're right. It's best to **just prohibit passing in nulls and throw explicit NPEs for both functions and consumers**. This, at least, can spare us from further debates :-)

mjsax replied:
SGTM. In general, even if possible, I don't like to give `null` semantics :)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,150 @@
+     */
+    public static <K, V> Branched<K, V> withFunction(
+            final Function<? super KStream<K, V>,
+                    ? extends KStream<K, V>> chain, final String name) {
```
nit: formatting
```
final Function<? super KStream<K, V>, ? extends KStream<K, V>> chain,
final String name) {
```

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,150 @@
+     * Create an instance of {@link Branched} with provided chain function and branch name postfix.
+     *
+     * @param chain A function that will be applied to the branch. If {@code null}, the identity
+     *              {@code kStream -> kStream} function will be supposed. If this function returns
```
used

If the provided function

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,150 @@
+    /**
+     * Create an instance of {@link Branched} with provided chain function and branch name postfix.
+     *
+     * @param chain A consumer to which the branch will be sent. If a non-null branch is provided here,
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ * Each branch (which is a {@link KStream} instance) then can be processed either by
+ * a {@link java.util.function.Function} or a {@link java.util.function.Consumer} provided via a {@link Branched}
+ * parameter. It also can be accessed from the {@link Map} returned by {@link BranchedKStream#defaultBranch(Branched)} or
+ * {@link BranchedKStream#noDefaultBranch()} (see <a href="#examples">usage examples</a>).
```
Maybe add a note that not all branches are added to the `Map`?

inponomarev replied:
Rephrased: `if certain conditions are met, it also can be accessed from the {@link Map}`... -- and all these coditions are described below

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ * <p>
+ * The branching happens on first-match: A record in the original stream is assigned to the corresponding result
+ * stream for the first predicate that evaluates to true, and is assigned to this stream only. If you need
+ * to route a record to multiple streams, you can use {@link KStream#filter(Predicate)} for each predicate instead
```
`you can apply [multiple] {#filter} [operators], one for each predicate, instead`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ * <h2><a name="maprules">Rules of forming the resulting map</a></h2>
+ * The keys of the {@code Map<String, KStream<K, V>>} entries returned by {@link BranchedKStream#defaultBranch(Branched)} or
+ * {@link BranchedKStream#noDefaultBranch()} are defined by the following rules:
+ * <p>
```
nit: Do we need a new paragraph? We get a list below anyway.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ * <ul>
+ *     <li>If {@link Named} parameter was provided for {@link KStream#split(Named)}, its value is used as
+ *     a prefix for each key. By default, no prefix is used
+ *     <li>If a name is provided for the {@link BranchedKStream#branch(Predicate, Branched)} via
```
`If a [branch?] name is provided [in] {@link BranchedKStream#branch(Predicate, Branched)} via the {@link Branched} parameter`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ * <p>
+ * <ul>
+ *     <li>If {@link Named} parameter was provided for {@link KStream#split(Named)}, its value is used as
+ *     a prefix for each key. By default, no prefix is used
```
nit: missing `.` at the end

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ *     <li>If {@link Named} parameter was provided for {@link KStream#split(Named)}, its value is used as
+ *     a prefix for each key. By default, no prefix is used
+ *     <li>If a name is provided for the {@link BranchedKStream#branch(Predicate, Branched)} via
+ *     {@link Branched} parameter, its value is appended to the prefix to form the {@code Map} key
```
nit: missing `.` at the end

inponomarev replied:
Do we use dots at the end of list items? (Sorry -- English is my second language so I really don't know. I noticed that sometimes you do but in most cases you don't. In Russian language we must always use dots or commas at the end of list items, but Russian and English punctuations are completely different.)

mjsax replied:
Also my second language... I think, it is mostly a matter of style. Personally, I prefer to use `.` at the end, because a bullet point is still a sentence from my POV, and actually, a bullet point could be multiple sentences. You also capitalized the first word of each bullet point, so it seems to be a sentence. -- I guess some people don't capitalize the first word and also don't use a `.` at the en, ie, us a "no-sentence" bullet point style. But this only work for "single sentence" bullet points.

It's really a nit. I just raised my personal (an obviously very subjective) preference.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ *     <li>If a name is provided for the {@link BranchedKStream#branch(Predicate, Branched)} via
+ *     {@link Branched} parameter, its value is appended to the prefix to form the {@code Map} key
+ *     <li>If a name is not provided for the branch, then the key defaults to {@code prefix + position} of the branch
+ *     as a decimal number, starting from {@code "1"}
```
nit: missing `.` at the end

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ *     <li>If a name is not provided for the {@link BranchedKStream#defaultBranch()} call, then the key defaults
+ *     to {@code prefix + "0"}
+ * </ul>
+ * <p>
```
nit: Do we need a new paragraph after a list?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ * </ul>
+ * <p>
+ * The values of the respective {@code Map<Stream, KStream<K, V>>} entries are formed as following:
+ * <p>
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ *     <li>If a chain function returns {@code null} for a given branch, then the respective entry is not put to the map.
+ *     <li>If a consumer is provided for a given branch, then the the respective entry is not put to the map
+ * </ul>
+ * <p>
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ * The values of the respective {@code Map<Stream, KStream<K, V>>} entries are formed as following:
+ * <p>
+ * <ul>
+ *     <li>If no chain function or consumer is provided {@link BranchedKStream#branch(Predicate, Branched)} via
```
`is provided [in]`

`via [the] {@link Branched} parameter.`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ * <ul>
+ *     <li>If no chain function or consumer is provided {@link BranchedKStream#branch(Predicate, Branched)} via
+ *     {@link Branched} parameter, then the value is the branch itself (which is equivalent to {@code ks -> ks}
+ *     identity chain function)
```
Do we need the content in the parentheses?

inponomarev replied:
No we don't, removed )

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ *     <li>If no chain function or consumer is provided {@link BranchedKStream#branch(Predicate, Branched)} via
+ *     {@link Branched} parameter, then the value is the branch itself (which is equivalent to {@code ks -> ks}
+ *     identity chain function)
+ *     <li>If a chain function is provided and returns a non-null value for a given branch, then the value is
```
`and returns` -> `that returns` (or `and it returns`)

`value` -> `{@link KStream}` ? (and `than the returned {@link KStream}` is added to the {@code Map}.`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ *     {@link Branched} parameter, then the value is the branch itself (which is equivalent to {@code ks -> ks}
+ *     identity chain function)
+ *     <li>If a chain function is provided and returns a non-null value for a given branch, then the value is
+ *     the result returned by this function
```
nit: missing "." (some for other bullet points)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ * <p>
+ * <ul>
+ *     <li>If no chain function or consumer is provided {@link BranchedKStream#branch(Predicate, Branched)} via
+ *     {@link Branched} parameter, then the value is the branch itself (which is equivalent to {@code ks -> ks}
```
`then the value is the branch itself` -> ` then the branch itself is added to the {@code Map}` ?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ *     identity chain function)
+ *     <li>If a chain function is provided and returns a non-null value for a given branch, then the value is
+ *     the result returned by this function
+ *     <li>If a chain function returns {@code null} for a given branch, then the respective entry is not put to the map.
```
`the respective entry is not put to the map.` -> `, then no entry is added to the map.`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ *     <li>If a chain function is provided and returns a non-null value for a given branch, then the value is
+ *     the result returned by this function
+ *     <li>If a chain function returns {@code null} for a given branch, then the respective entry is not put to the map.
+ *     <li>If a consumer is provided for a given branch, then the the respective entry is not put to the map
```
`If a [non-null] consumer`

`then the the respective entry is not put to the map` -> `, then no entry is added to the map.`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ *     the result returned by this function
+ *     <li>If a chain function returns {@code null} for a given branch, then the respective entry is not put to the map.
+ *     <li>If a consumer is provided for a given branch, then the the respective entry is not put to the map
+ * </ul>
```
Seems the `null` consumer case is missing (is case we want to allow if, but as mentioned above, I have some doubts)?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ * <pre> {@code
+ * source.split()
+ *     .branch((key, value) -> value.contains("A"), Branched.withConsumer(ks -> ks.to("A")))
+ *     .branch((key, value) -> value.contains("B"), Branched.withConsumer(ks -> ks.to("B")))
```
To simplify the example, should we use `predicate1`, `predicate2` as in the first example?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+     * Finalizes the construction of branches and defines the default branch for the messages not intercepted
+     * by other branches.
+     *
+     * @return {@link Map} of named branches. For rules of forming the resulting map, see {@link BranchedKStream}
```
`{@code BranchedKStream}`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,172 @@
+ * The branching happens on first-match: A record in the original stream is assigned to the corresponding result
+ * stream for the first predicate that evaluates to true, and is assigned to this stream only. If you need
+ * to route a record to multiple streams, you can use {@link KStream#filter(Predicate)} for each predicate instead
+ * of branching.
```
Add: `If a record does not match any predicates, it will be routed to the default branch, or dropped if no default branch is created.`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -773,10 +775,32 @@
     KStream<K, V>[] branch(final Named named, final Predicate<? super K, ? super V>... predicates);
 
+    /**
+     * Splits this stream. {@link BranchedKStream} can be used for routing the records to different branches depending
```
Nit `Split` (no `s`) -- we use imperative to write JavaDocs.

`this stream` -> `this {@code KStream}.`

(Same for the overload method)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/BranchedInternal.java:
```diff
@@ -0,0 +1,62 @@
+        }
+    }
+
+    public void process(final KStreamImpl<K, V> newStream, final String branchChildName, final Map<String, KStream<K, V>> result) {
```
`newStream` -> `branch` ?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/BranchedInternal.java:
```diff
@@ -0,0 +1,62 @@
+        } else {
+            result.put(branchChildName, transformedStream);
+        }
+    }
```
I think this method is hard to read. Proposal:
```
if (chainFunction != null) {
  final KStream<K, V> transformedStream = chainFunction.apply(newStream);
  if (transformedStream != null) {
    result.put(branchChildName, transformedStream);
  }
} else if (chainConsumer != null) {
  chainConsumer.accept(transformedStream);
} else {
  result.put(branchChildName, newStream);
}
```

inponomarev replied:
Agreed, replaced the code as you suggested

mjsax commented on streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:
```diff
@@ -615,7 +615,7 @@ public void shouldUseSpecifiedNameForTransformValuesWithKey() {
 
     @Test
-    @SuppressWarnings("unchecked")
+    @SuppressWarnings({"unchecked", "deprecation"})
```
Seems we should add a corresponding test for `split()` ? (Same for other test classes.)

inponomarev replied:
done, added a new corresponding test!

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -492,6 +496,16 @@ public void foreach(final ForeachAction<? super K, ? super V> action,
+
+    @Override
+    public BranchedKStream<K, V> split(final Named named) {
+        return new BranchedKStreamImpl<>(this, repartitionRequired, new NamedInternal(named));
```
missing `null` check for `named`

inponomarev replied:
done

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamBranch.java:
```diff
@@ -41,14 +43,18 @@
+                    return;
                 }
             }
+            // using default child node if supplied
```
Should we really do it this way, or add a predicate to the list that always returns `true` if a default branch is added?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/BranchedKStreamImpl.java:
```diff
@@ -0,0 +1,97 @@
+
+    @Override
+    public BranchedKStream<K, V> branch(final Predicate<? super K, ? super V> predicate) {
+        return branch(predicate, BranchedInternal.empty());
```
Missing `null` check for `predicate` (similar below for other methods and parameters).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/BranchedKStreamImpl.java:
```diff
@@ -0,0 +1,97 @@
+
+    @Override
+    public Map<String, KStream<K, V>> defaultBranch(final Branched<K, V> branched) {
+        createBranch(branched, 0);
```
Should we call `branch((k,v) -> true, branched)` instead to just add a predicate and branch? This way, the default branch is nothing special at runtime any longer.

inponomarev replied:
The default branch should have index 0 (so it will be stable when branches are added or removed), but it should always be checked after all other branches. And when we come to the default branch during message processing, there is actually no need in dereferncing a predicate and calling `test`... that's why I treat the default branch differently.

mjsax replied:
I guess it's fine both ways. -- The point about the index is a good one that I missed. But would still be doable I guess.

I don't think that there would be any measurable runtime difference if you use a "default predicate" (what we also do in the current implementation) -- the code is just a little "cleaner" as we don't need an extra "if" at the end -- but it's also not the end of the world as the `process` method is fairly simply anyway.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/BranchedKStreamImpl.java:
```diff
@@ -0,0 +1,97 @@
+
+    private void createBranch(final Branched<K, V> branched, final int index) {
+        final BranchedInternal<K, V> branchedInternal = new BranchedInternal<>(branched);
+        final String branchChildName = branchedInternal.branchProcessorName(splitterName, index);
```
As we call `branchProcessorName` only once, I am wondering if we should embed the code here and remove `branchProcessorName` -- our usual pattern is to only have getters on the `XxxInternal` implemenation.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/BranchedKStreamImpl.java:
```diff
@@ -0,0 +1,97 @@
+    private final KStreamImpl<K, V> source;
+    private final boolean repartitionRequired;
+    private final String splitterName;
+    private final Map<String, KStream<K, V>> result = new HashMap<>();
```
`result` -> `outputBranches` ?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/BranchedKStreamImpl.java:
```diff
@@ -0,0 +1,97 @@
+        this.repartitionRequired = repartitionRequired;
+        this.splitterName = named.orElseGenerateWithPrefix(source.builder, BRANCH_NAME);
+
+        // predicates and childNames are passed by reference so when the user adds a branch they get added to
```
I am wondering if it might be better to move this code into a `build` method that would be called within `defaultBranch()` / `noDefaultBranch()` ?

The pattern to pass in empty list that we modify later seems undesirable, and we should first build the list, and than pass them in -- otherwise, we make assumptions how `ProcessorParameters` and `ProcessorGraphNode` might be implemented what we should avoid.

inponomarev replied:
I clearly remember that something made me to write it this way, but I have to recall...

mjsax replied:
Would love to learn about it. -- In general, it's easier to follow the same pattern throughout the code base. It easier to reason about the code that way, and also easier for people to learn the code base.

mjsax replied:
Just saw your other comment: https://github.com/apache/kafka/pull/9107#issuecomment-751261181

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/BranchedKStreamImpl.java:
```diff
@@ -0,0 +1,97 @@
+        final KStreamImpl<K, V> newStream = new KStreamImpl<>(branchChildName, source.keySerde,
+                source.valueSerde, source.subTopologySourceNodes,
+                repartitionRequired, branchChildNode, source.builder);
+        branchedInternal.process(newStream, branchChildName, result);
```
Same comment as for `branchProcessorName` -- it seems better to keep `BranchedInternal` a simple container and pull in the code into this method.

inponomarev replied:
done

mjsax reviewed (commented):
Thanks for the PR @inponomarev and sorry for the long wait for a review

Many comments are about JavaDocs, so it's mostly small suggestions. A few comments about the code structure are there, too.

inponomarev committed:
code review fixes

inponomarev committed:
requireNonNull for chain function and consumer

inponomarev commented:
Hi @mjsax, thanks for your thorough revew! I have fixed everything according to your comments, except:

* https://github.com/apache/kafka/pull/9107#discussion_r547592368 -- just want to be sure!
* https://github.com/apache/kafka/pull/9107#discussion_r547606064 -- see my explanation
* https://github.com/apache/kafka/pull/9107#discussion_r547607739 -- I agree that this is ugly but there was a good reason why I made it this way. Just give me a day and I'll return with comments (or will fix it).

inponomarev committed:
test for branching with no defaultBranch()/noDefaultBranch()

inponomarev commented:
OK @mjsax concerning https://github.com/apache/kafka/pull/9107#discussion_r547607739 I remembered why it was implemented this way!

The problem is that **it is not necessary to invoke `defaultBranch()` / `noDefaultBranch()`** when we use consumers, like in this simple example (I just added a new unit test for this case):
 
```java
source.split()
    .branch(isCoffee, Branched.withConsumer(issuer::setCoffeePurchases))
    .branch(isElectronics, Branched.withConsumer(issuer::setElectronicsPurchases));
```

mjsax commented:
About the original comment: https://github.com/apache/kafka/pull/9107#issuecomment-666749809

I am fine with those changes.

About https://github.com/apache/kafka/pull/9107#issuecomment-751261181 -- that is a good point. Thanks for explaining. I guess it's a "philosophical" question if we want to allow this pattern though, or if we want to require that either `defaultBranch()` or `noDefaultBranch()` is called? -- I did consider calling `branch()` like a builder pattern, and the final `[noD|d]efaultBranch` call is basically `build()`?

Curious to hear what @vvcephei thinks about it.

mjsax commented:
@inponomarev -- Can you also update the docs for Kafka Streams and the 2.8 upgrade guide in this PR.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/BranchedInternal.java:
```diff
@@ -0,0 +1,49 @@
+        return new BranchedInternal<>();
+    }
+
+    String getName() {
```
It's a naming convention in the whole Kafka code base, to omit the `get` prefix for all getter methods, ie, this should be `name()`. (Similar below for the other getters.)

inponomarev replied:
done

inponomarev committed:
changelog, naming convention for BranchedInternal

inponomarev commented:
> Can you also update the docs for Kafka Streams and the 2.8 upgrade guide in this PR.

The documentation had been already updated (see changes in `docs/streams/developer-guide/dsl-api.html`)

I also modified  `docs/upgrade.html` -- should I add something more here, like code examples?

Another question: CI checks fail because of usage of deprecated `branch` method in `streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala`. Since I'm not a Scala user, I have no idea of what should be done here. 

Most likely we should deprecate the `branch` method and add a wrapper for the new `split` method, but I don't know how to do this correctly.

mjsax commented:
To make the build pass, for now, it should be sufficient to just deprecate the method via `@nowarn("cat=deprecation")` -- But it seems we should update the Scala API, too. If you cannot handle it, we can do a follow up PR.

It seems, we need to add `split()` to `KStream.scala` and introduce a new `BranchedKStream.scala` and `Branch.scala` classes and maybe some translations from Java `Consumer`/`Function` to their Scala variants. But I also don't really now Scala; @vvcephei should nkow better.

inponomarev commented:
As far as I can judge from the name, `@nowarn` is not for deprecation, but rather for a warning suppression  🤔 apparently we need to mirror the changes in Java `KStream` interface here. Never wrote anything in Scala before. OK, it's better to wait for @vvcephei !

mjsax commented:
Maybe I miss-understood you question. I thought the build fails because we are using some deprecated method -- for this case, we can make the build pass by suppressing the warning. If you want to deprecate a method in the Scala API, you just add `@deprecated` similar to Java. -- I guess it makes sense to also deprecate the `KStream.scala#branch()` method, but suppressing the warning should also make the build pass and we can deprecate this method when we add the new `split()` method.

vvcephei commented:
Hey @inponomarev and @mjsax ! I'm glad to see this is moving along.

Regarding https://github.com/apache/kafka/pull/9107#issuecomment-751922810 :

My understanding was that defaultBranch/noDefaultBranch were the terminal operators, in that they close out the context of a BranchedKStream, and you can't add any more branches after one of those methods.

But also, the whole branching construct is an incremental builder like the rest of the Kafka Streams API. In other words, just like this is a valid program:

```java
builder.stream("input")
       .filter(myPredicate)
```

so would be Ivan's example:

```java
builder.split()
       .branch("myBranch", ...)
```

What I mean by "incremental builder" is that each time you call a chained method in the DSL, it immediately adds nodes to the program, as opposed to having to call any kind of `build()` method to actually add stuff to the program. I think there are pros and cons to this design, but it seems more in line with the rest of the DSL not to require the terminal operators.

inponomarev commented:
Hi @vvcephei , thank you for your comment. There's another question that we were unable to solve without you -- see https://github.com/apache/kafka/pull/9107#issuecomment-752098770 from the words 'CI checks fail' and further discussion. Can you clarify, what's expected from `KStream.scala` ?

mjsax commented:
@vvcephei -- hope you are also ok with the proposed changes to the KIP as per the PR description on top: https://github.com/apache/kafka/pull/9107#issuecomment-666749809

inponomarev committed:
Scala API wrapper

inponomarev commented:
@vvcephei @mjsax I added full Scala wrapper for the new API: `split` method, `BranchedKStream` and `Branched`. Also added Scala unit tests that verify main use cases

Commit:
spotlessApply

Commit:
Correct the test to use the Scala Branched, not the Java one.

Commit:
Use default arguments instead of method overloads

Commit:
Use "suffix" instead of "postfix"

Commit:
convert CRLF -> LF

vvcephei commented:
Hey @inponomarev , I just took a look at the Scala API. Thanks for adding that!

I figured it'd be just easier to push a few tweaks than to describe what needs to be done.

* You asked me offline if we could avoid the overloads in Branched, and indeed, we can with a default argument of `null` for the name.
* The Scala test was inadvertently using the Java Branched class, but you meant to test the Scala one.
* I happened to notice a small typo: The opposite of "prefix" is "suffix", not "postfix"
* I also noticed that your files all had CRLF (windows) return characters, so I fixed them. You might want to configure git for `autocrlf` (`git config --global core.autocrlf true`) (see https://www.git-scm.com/book/en/v2/Customizing-Git-Git-Configuration)

These are all separate commits above, so you can scrutinize each one. This PR is your work, so feel free to protest any of my suggestions.

inponomarev committed:
fixed JavaDoc/Scaladoc

* consumers cannot be null
* typo: "function"->"consumer"

inponomarev commented:
Hi @vvcephei thank you for your commits! Is everything else OK, especially https://github.com/apache/kafka/pull/9107#issuecomment-666749809?

@mjsax I pushed small fixes to Javadoc/Scaladoc, and AFAICS only tests not related to the changes are failing.

inponomarev commented:
Thank you @vvcephei, I have updated [the KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-418%3A+A+method-chaining+way+to+branch+KStream) and now it reflects the actual implementation.

I just wasn't sure if it's ok to edit specification text after it has been formally approved :-)

inponomarev commented:
Ooops, I just noticed one more thing to fix: https://github.com/apache/kafka/blob/3b7dda459ca9a1bcea72ca981c2a19a4ccffe9f2/streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphTest.java#L115

Will push a new commit today or tomorrow!

inponomarev committed:
Rewrite StreamsGraphTest using split()

vvcephei commented:
Thanks, @inponomarev !

Yes, we sometimes have to update the KIP as the PR gets finalized. I like to send a quick reply to the vote thread on the mailing list to summarize the design changes, in case anyone wants to object (I don't think I've ever had someone object).

mjsax commented on docs/streams/developer-guide/dsl-api.html:
```diff
@@ -366,24 +366,28 @@ <h4 class="anchor-heading"><a id="streams_concepts_globalktable" class="anchor-l
+.branch((key, value) -> key.startsWith("B"),  /* second predicate */
+Branched.as("B"))  
+.defaultBranch(Branched.as("C"))
+);
```
nit: seems we use some indention for this code block?
```
Map<String, KStream<String, Long>> branches =
    stream.split(Named.as("Branch-"))
          .branch((key, value) -> key.startsWith("A"),  /* first predicate  */ 
                  Branched.as("A")) 
          .branch((key, value) -> key.startsWith("B"),  /* second predicate */
                  Branched.as("B"))  
        .defaultBranch(Branched.as("C"))
);
```

mjsax commented on docs/upgrade.html:
```diff
@@ -21,6 +21,12 @@
 <h5><a id="upgrade_280_notable" href="#upgrade_280_notable">Notable changes in 2.8.0</a></h5>
 
+<ul>
+    <li>Kafka Streams introduce a type-safe split() operator as a substitution for deprecated KStream#branch method
```
nit: `split()` -> `<code>split()</code>`
Same for `KStream#branch` (should we add `()`, too?)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,149 @@
+     */
+    public static <K, V> Branched<K, V> withFunction(
+            final Function<? super KStream<K, V>, ? extends KStream<K, V>> chain) {
+        Objects.requireNonNull(chain, "chain function should not be null");
```
nit: `should not` -> `cannot`  (or `can't` if you prefer)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,149 @@
+     * @return a new instance of {@code Branched}
+     */
+    public static <K, V> Branched<K, V> withConsumer(final Consumer<KStream<K, V>> chain) {
+        Objects.requireNonNull(chain, "chain consumer should not be null");
```
nit: `should not` -> `cannot`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,149 @@
+    /**
+     * Create an instance of {@code Branched} with provided chain consumer.
+     *
+     * @param chain A consumer to which the branch will be sent. If a consumer is provided here,
```
nit: remove `here`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,149 @@
+    /**
+     * Create an instance of {@code Branched} with provided chain function and branch name suffix.
+     *
+     * @param chain A function that will be applied to the branch. If {@code null}, the identity
```
remove ` If {@code null}, the identity {@code kStream -> kStream} function will be used.` -- we don't allow `null`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,149 @@
+     */
+    public static <K, V> Branched<K, V> withFunction(
+            final Function<? super KStream<K, V>, ? extends KStream<K, V>> chain, final String name) {
+        Objects.requireNonNull(chain, "chain function should not be null");
```
nit: `should not` -> `cannot`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,149 @@
+    /**
+     * Create an instance of {@code Branched} with provided chain consumer and branch name suffix.
+     *
+     * @param chain A consumer to which the branch will be sent. If a non-null consumer is provided here,
```
nit: remove `here`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,149 @@
+     */
+    public static <K, V> Branched<K, V> withConsumer(final Consumer<? super KStream<K, V>> chain,
+                                                     final String name) {
+        Objects.requireNonNull(chain, "chain consumer should not be null");
```
nit: `should not` -> `cannot`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+ * (see <a href="#examples">usage examples</a>).
+ * <p>
+ * The branching happens on first-match: A record in the original stream is assigned to the corresponding result
+ * stream for the first predicate that evaluates to true, and is assigned to this stream only. If you need
```
nit: `{@code true}`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+ * {@link BranchedKStream#defaultBranch(Branched)} or {@link BranchedKStream#noDefaultBranch()}
+ * (see <a href="#examples">usage examples</a>).
+ * <p>
+ * The branching happens on first-match: A record in the original stream is assigned to the corresponding result
```
nit: `first match` (without dash) ?

mjsax replied:
\cc @vvcephei

vvcephei replied:
Yes, I agree, unless you want to add a noun:

```suggestion
 * The branching happens on a first-match basis: A record in the original stream is assigned to the corresponding result
```

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+ * <p>
+ * The branching happens on first-match: A record in the original stream is assigned to the corresponding result
+ * stream for the first predicate that evaluates to true, and is assigned to this stream only. If you need
+ * to route a record to multiple streams, you can apply multiple {@link KStream#filter(Predicate)} operators,
```
proposal (add) `operators[ to the same {@link KStream} instance],`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+ * one for each predicate, instead of branching.
+ * <p>
+ * The process of routing the records to different branches is a stateless record-by-record operation.
+ * <h2><a name="maprules">Rules of forming the resulting map</a></h2>
```
nit: add empty line above headline for better readability in the code (won't change the produced html).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+ * <ul>
+ *     <li>If no chain function or consumer is provided in {@link BranchedKStream#branch(Predicate, Branched)} via
+ *     the {@link Branched} parameter, then the the branch itself is added to the {@code Map}
+ *     <li>If a non-null chain function is provided and it returns a non-null value for a given branch, then the value
```
`If a non-null chain function` -> `If a chain function`

(We don't allow `null` as chain function.)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+ *     <li>If a non-null chain function is provided and it returns a non-null value for a given branch, then the value
+ *     is the result returned by this function
+ *     <li>If a chain function returns {@code null} for a given branch, then no entry is added to the map
+ *     <li>If a non-null consumer is provided for a given branch, then no entry is added to the map
```
Remove `non-null` (`null` is not allowed anyway)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+ *     the {@link Branched} parameter, then the the branch itself is added to the {@code Map}
+ *     <li>If a non-null chain function is provided and it returns a non-null value for a given branch, then the value
+ *     is the result returned by this function
+ *     <li>If a chain function returns {@code null} for a given branch, then no entry is added to the map
```
Should we merge this sentence to the previous bullet point, instead of making a new bullet point?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+ * Each branch (which is a {@link KStream} instance) then can be processed either by
+ * a {@link java.util.function.Function} or a {@link java.util.function.Consumer} provided via a {@link Branched}
+ * parameter. If certain conditions are met, it also can be accessed from the {@link Map} returned by
+ * {@link BranchedKStream#defaultBranch(Branched)} or {@link BranchedKStream#noDefaultBranch()}
```
I think it's worth to highlight that `defaultBranch / noDefaultBranch` are optional (to avoid the confusion about "builder pattern"):

`by an optional {@link BranchedKStream#defaultBranch(Branched)} or {@link BranchedKStream#noDefaultBranch()}  method call (see ...).`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+ */
+public interface BranchedKStream<K, V> {
+    /**
+     * Defines a branch for records that match the predicate.
```
`Define` (no `s` -- we use imperative in JavaDocs)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+    BranchedKStream<K, V> branch(Predicate<? super K, ? super V> predicate);
+
+    /**
+     * Defines a branch for records that match the predicate.
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+    BranchedKStream<K, V> branch(Predicate<? super K, ? super V> predicate, Branched<K, V> branched);
+
+    /**
+     * Finalizes the construction of branches and defines the default branch for the messages not intercepted
```
`Finalize` (no `s')

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+
+    /**
+     * Finalizes the construction of branches and defines the default branch for the messages not intercepted
+     * by other branches.
```
Add:
```
Calling {@code defaultBranch()} or {@link #noDefaultBranch()} is optional.
```

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+    Map<String, KStream<K, V>> defaultBranch();
+
+    /**
+     * Finalizes the construction of branches and defines the default branch for the messages not intercepted
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+
+    /**
+     * Finalizes the construction of branches and defines the default branch for the messages not intercepted
+     * by other branches.
```
Add:
```
Calling {@code defaultBranch()} or {@link #noDefaultBranch()} is optional.
```

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/BranchedKStream.java:
```diff
@@ -0,0 +1,169 @@
+    Map<String, KStream<K, V>> defaultBranch(Branched<K, V> branched);
+
+    /**
+     * Finalizes the construction of branches without forming a default branch.
```
as above

mjsax replied:
Add:
```
Calling {@code noDefaultBranch()} or {@link #defaultBranch()} is optional.
```

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -773,10 +775,32 @@
      * @param named  a {@link Named} config used to name the processor in the topology
      * @param predicates the ordered list of {@link Predicate} instances
      * @return multiple distinct substreams of this {@code KStream}
+     * @deprecated since 2.7. Use {@link #split(Named)} instead.
```
`2.7` -> `2.8`

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Branched.scala:
```diff
@@ -0,0 +1,75 @@
+  /**
+   * Create an instance of `Branched` with provided chain function and branch name suffix.
+   *
+   * @param chain A function that will be applied to the branch. If `null`, the identity
```
Remove `If null, the identity function will be used.` (we don't allow `null`)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,149 @@
+     * @return a new instance of {@code Branched}
+     */
+    public static <K, V> Branched<K, V> as(final String name) {
+        return new Branched<>(name, null, null);
```
Missing `null` check for `name`

mjsax replied:
\cc @vvcephei

vvcephei replied:
I agree, it seems like a good idea to check for `null` here.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java:
```diff
@@ -0,0 +1,149 @@
+    /**
+     * Create an instance of {@code Branched} with provided branch name suffix.
+     *
+     * @param name the branch name suffix to be used. If {@code null}, a default branch name suffix will be generated
```
remove: `If {@code null}, a default branch name suffix will be generated` (we should not allow `null` to align to other methods, eg, `Named.as()`)

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Branched.scala:
```diff
@@ -0,0 +1,75 @@
+   *              `null`, its result is ignored, otherwise it is added to the Map returned
+   *              by [[BranchedKStream.defaultBranch]] or [[BranchedKStream.noDefaultBranch]] (see
+   *              [[BranchedKStream]] description for details).
+   * @param name  the branch name suffix to be used. If `null`, a default branch name suffix will be generated
```
remove `If null, a default branch name suffix will be generated` -- `null` should not be allowed (cf my comment on the corresponding Java class)

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Branched.scala:
```diff
@@ -0,0 +1,75 @@
+   *              the respective branch will not be added to the resulting Map returned
+   *              by [[BranchedKStream.defaultBranch]] or [[BranchedKStream.noDefaultBranch]] (see
+   *              [[BranchedKStream]] description for details).
+   * @param name  the branch name suffix to be used. If `null`, a default branch name suffix will be generated
```
as above; remove `If null, a default branch name suffix will be generated`

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/BranchedKStream.scala:
```diff
@@ -0,0 +1,116 @@
+ *
+ * Each branch (which is a [[KStream]] instance) then can be processed either by
+ * a function or a consumer provided via a [[Branched]]
+ * parameter. If certain conditions are met, it also can be accessed from the Map returned by
```
nit:
```
Map -> `Map`
```

add:
```
by [an optional] defaultBranch or noDefaultBranch [method call].
```

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/BranchedKStream.scala:
```diff
@@ -0,0 +1,116 @@
+ * parameter. If certain conditions are met, it also can be accessed from the Map returned by
+ * [[defaultBranch]] or [[noDefaultBranch]].
+ * <p>
+ * The branching happens on first-match: A record in the original stream is assigned to the corresponding result
```
`first match` (no dash?)

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/BranchedKStream.scala:
```diff
@@ -0,0 +1,116 @@
+ * [[defaultBranch]] or [[noDefaultBranch]].
+ * <p>
+ * The branching happens on first-match: A record in the original stream is assigned to the corresponding result
+ * stream for the first predicate that evaluates to true, and is assigned to this stream only. If you need
```
nit:
```
true -> `true`
```

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/BranchedKStream.scala:
```diff
@@ -0,0 +1,116 @@
+ * <p>
+ * The branching happens on first-match: A record in the original stream is assigned to the corresponding result
+ * stream for the first predicate that evaluates to true, and is assigned to this stream only. If you need
+ * to route a record to multiple streams, you can apply multiple [[KStream.filter]] operators,
```
nit:
`operators[ to the same [[KStream]] instance],`

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/BranchedKStream.scala:
```diff
@@ -0,0 +1,116 @@
+class BranchedKStream[K, V](val inner: BranchedKStreamJ[K, V]) {
+
+  /**
+   * Defines a branch for records that match the predicate.
```
nit: `Define` (no `s`)

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/BranchedKStream.scala:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  /**
+   * Defines a branch for records that match the predicate.
```
nit: `Define` (no `s`)

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/BranchedKStream.scala:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  /**
+   * Finalizes the construction of branches and defines the default branch for the messages not intercepted
```
nit: `Finalize` (no `s`)

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/BranchedKStream.scala:
```diff
@@ -0,0 +1,116 @@
+
+  /**
+   * Finalizes the construction of branches and defines the default branch for the messages not intercepted
+   * by other branches.
```
Add: `Calling [[defaultBranch]] or [[noDefaultBranch]] is optional.`

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/BranchedKStream.scala:
```diff
@@ -0,0 +1,116 @@
+
+  /**
+   * Finalizes the construction of branches and defines the default branch for the messages not intercepted
+   * by other branches.
```
Add: `Calling [[defaultBranch]] or [[noDefaultBranch]] is optional.`

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/BranchedKStream.scala:
```diff
@@ -0,0 +1,116 @@
+  def defaultBranch(): Map[String, KStream[K, V]] = toScalaMap(inner.defaultBranch())
+
+  /**
+   * Finalizes the construction of branches and defines the default branch for the messages not intercepted
```
nit: `Finalize` (no `s`)

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/BranchedKStream.scala:
```diff
@@ -0,0 +1,116 @@
+  def defaultBranch(branched: Branched[K, V]): Map[String, KStream[K, V]] = toScalaMap(inner.defaultBranch(branched))
+
+  /**
+   * Finalizes the construction of branches without forming a default branch.
```
nit: `Finalize` (no `s`)

mjsax replied:
Add: `Calling [[noDefaultBranch]] or [[defaultBranch]] is optional.`

mjsax reviewed (commented):
Sorry for late review... Overall LGTM. Just nits.

mjsax commented:
@inponomarev the failing tests seems to be due to a known issue that was fixed via https://github.com/apache/kafka/pull/9768

Can you rebase your PR to pickup the fix so we can get a green build?

inponomarev commented:
> @inponomarev the failing tests seems to be due to a known issue that was fixed via #9768
> 
> Can you rebase your PR to pickup the fix so we can get a green build?

Done rebasing, expect the fixes according to your latest review soon!

inponomarev committed:
Javadoc/Scaladoc and Developer Guide updates

inponomarev commented:
Hi @mjsax , I have implemented all your suggestions concerning Javadocs/Scaladocs.

After rebasing, there are less failing tests

I also see that `JDK8` build continues to fail because of Scala compilation issue. AFAICS, that's because we are using `FunctionWrappers` object which is not available in older versions of Scala. Is it OK?

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Branched.scala:
```diff
@@ -0,0 +1,73 @@
+
+import org.apache.kafka.streams.kstream.{Branched => BranchedJ, KStream => KStreamJ}
+
+import scala.jdk.FunctionConverters._
```
Based on your comment, you are right, that those `FunctionConverters ` are only available in Scala 2.13, but the code must compile with Scala 2.12.

Thus we must remove this import and do the `Function/Consumer` conversion below manually (cf below).

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Branched.scala:
```diff
@@ -0,0 +1,73 @@
+  def withFunction[K, V](chain: KStream[K, V] => KStream[K, V], name: String = null): BranchedJ[K, V] =
+    BranchedJ.withFunction({ s: KStreamJ[K, V] =>
+      chain.apply(new KStream[K, V](s)).inner
+    }.asJava, name)
```
Instead of using `asJava` we can do those three lines as:
```
    BranchedJ.withFunction((f: KStreamJ[K, V]) => chain.apply(new KStream[K, V](f)).inner, name)
```

inponomarev replied:
Hmmm, this looks prettier :-)

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Branched.scala:
```diff
@@ -0,0 +1,73 @@
+  def withConsumer[K, V](chain: KStream[K, V] => Unit, name: String = null): BranchedJ[K, V] =
+    BranchedJ.withConsumer({ s: KStreamJ[K, V] =>
+      chain.apply(new KStream[K, V](s))
+    }.asJava, name)
```
Instead of using `asJava` we can do those three lines as:
```
    BranchedJ.withConsumer((c: KStreamJ[K, V]) => chain.apply(new KStream[K, V](c)), name)
```

mjsax reviewed (commented):
@inponomarev -- I believe you are right, and we need to address it, as the code must compile with Scala 2.12. I am not a Scala person so take my inline comments with a grain of salt, but it did compile for me locally using Scala 2.12 (ie, `./gradlew -PscalaVersion=2.12 scale systemTestLibs`) and Scala 2.13, after I applied the suggested changes while I saw Scala 2.12 fail before.

You will also need to rebase to resolve merge conflicts.

\cc @vvcephei to confirm that the proposal looks right. Not sure to what extend we would need to mimic the super/sub-type definition from Java (ie, `Function<? super KStream<K, V>, ? extends KStream<K, V>>`) -- but IIRC for previous PR, this was not required in Scala?

inponomarev committed:
unused Named import

inponomarev committed:
remove FunctionConverters

inponomarev commented:
Hi @mjsax , I have rebased and manually merged conflicts, and also removed `FunctionConverters`

JDK8 build still fails, but this time much later -- something related to integration testing

mjsax commented:
Wait failure do you see? Seem Jenkins in still running.

inponomarev commented:
I was talking about build 17 (triggered by Commit db573f5, see https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-9107/)

Where did build 18 came from, why it took 8 hours and then timed out -- I can't understand 😃

mjsax commented:
Ah I see -- well, we do have from flaky tests, so nothing to worry about I guess. The last run timed out, so I retriggered the build. However, I could build it locally with Java8/Scale 2.12 and so I guess we can merge. Just waiting for @vvcephei to take a quick look at the last Scala commit.

vvcephei commented:
Thanks, all, that Scala fix looks perfect to me.

mjsax commented:
Merged to `trunk`.

Congrats for getting this into the 2.8.0 release @inponomarev -- great work!
